File: javascript/update_docs_db.js
#!/usr/bin/env node
// update_docs_db.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:scripts/update_docs_db.py
// Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

/**
 * Scannt alle vorhandenen Dokumentationen und aktualisiert docs.db
 */

const fs = require('fs');
const path = require('path');
const sqlite3 = require('sqlite3').verbose();

const WORKSPACE = '/home/openclaw/.openclaw/workspace';
const DB_PATH = path.join(WORKSPACE, 'db', 'docs.db');

function scanDocumentations() {
    /**Scannt alle .md Dateien im Workspace*/
    const docs = [];
    
    // Hauptverzeichnis
    const mainFiles = fs.readdirSync(WORKSPACE)
        .filter(file => file.endsWith('.md'))
        .map(file => path.join(WORKSPACE, file))
        .filter(filePath => fs.statSync(filePath).isFile() && !fs.lstatSync(filePath).isSymbolicLink());
    
    for (const filePath of mainFiles) {
        const fileName = path.basename(filePath);
        docs.push({
            name: fileName,
            path: '/',
            category: 'main',
            description: getDescription(filePath),
            type: 'doc',
            has_symlink: false,
            symlink_path: null,
            last_update: getMtime(filePath)
        });
    }
    
    // WebSearch Verzeichnis
    const websearchDir = path.join(WORKSPACE, 'websearch');
    if (fs.existsSync(websearchDir)) {
        const websearchFiles = fs.readdirSync(websearchDir)
            .filter(file => file.endsWith('.md'))
            .map(file => path.join(websearchDir, file));
        
        for (const filePath of websearchFiles) {
            const fileName = path.basename(filePath);
            docs.push({
                name: fileName,
                path: 'websearch/',
                category: 'websearch',
                description: getDescription(filePath),
                type: fileName.includes('GUIDE') ? 'guide' : 'config',
                has_symlink: true,
                symlink_path: `websearch/${fileName}`,
                last_update: getMtime(filePath)
            });
        }
    }
    
    // MCP Verzeichnis
    const mcpDir = path.join(WORKSPACE, 'mcp');
    if (fs.existsSync(mcpDir)) {
        const mcpFiles = fs.readdirSync(mcpDir)
            .filter(file => file.endsWith('.md'))
            .map(file => path.join(mcpDir, file));
        
        for (const filePath of mcpFiles) {
            const fileName = path.basename(filePath);
            const isSymlink = fs.lstatSync(filePath).isSymbolicLink();
            docs.push({
                name: fileName,
                path: 'mcp/',
                category: 'mcp',
                description: getDescription(filePath),
                type: isSymlink ? 'symlink' : 'guide',
                has_symlink: isSymlink,
                symlink_path: isSymlink ? fs.readlinkSync(filePath) : null,
                last_update: getMtime(filePath)
            });
        }
    }
    
    // Docs-Unterverzeichnisse
    const docsDir = path.join(WORKSPACE, 'docs');
    if (fs.existsSync(docsDir)) {
        const subdirs = fs.readdirSync(docsDir)
            .filter(item => fs.statSync(path.join(docsDir, item)).isDirectory());
        
        for (const subdir of subdirs) {
            const subdirPath = path.join(docsDir, subdir);
            const subdirFiles = fs.readdirSync(subdirPath)
                .filter(file => file.endsWith('.md'))
                .map(file => path.join(subdirPath, file));
            
            for (const filePath of subdirFiles) {
                const fileName = path.basename(filePath);
                docs.push({
                    name: fileName,
                    path: `docs/${subdir}/`,
                    category: subdir,
                    description: getDescription(filePath),
                    type: 'doc',
                    has_symlink: false,
                    symlink_path: null,
                    last_update: getMtime(filePath)
                });
            }
        }
    }
    
    // Cluster, Memory, Reports, Skills
    const categories = ['cluster', 'memory', 'reports', 'skills'];
    for (const category of categories) {
        const catDir = path.join(WORKSPACE, category);
        if (fs.existsSync(catDir)) {
            const catFiles = fs.readdirSync(catDir)
                .filter(file => file.endsWith('.md'))
                .map(file => path.join(catDir, file));
            
            for (const filePath of catFiles) {
                const fileName = path.basename(filePath);
                docs.push({
                    name: fileName,
                    path: `${category}/`,
                    category: category,
                    description: getDescription(filePath),
                    type: 'doc',
                    has_symlink: false,
                    symlink_path: null,
                    last_update: getMtime(filePath)
                });
            }
        }
    }
    
    return docs;
}

function getDescription(filePath) {
    /**Extrahiert erste Zeile als Beschreibung*/
    try {
        const content = fs.readFileSync(filePath, 'utf8');
        const firstLine = content.split('\n')[0].trim();
        if (firstLine.startsWith('#')) {
            return firstLine.replace(/^#+\s*/, '');
        }
        return firstLine.length > 50 ? firstLine.substring(0, 50) + '...' : firstLine;
    } catch (error) {
        return 'Dokumentation';
    }
}

function getMtime(filePath) {
    /**Gibt letzte Änderung zurück*/
    try {
        const stats = fs.statSync(filePath);
        const date = new Date(stats.mtime);
        return date.toISOString().split('T')[0];
    } catch (error) {
        return '2026-04-18';
    }
}

function updateDatabase(docs) {
    /**Aktualisiert docs.db mit allen gefundenen Dokumenten*/
    const db = new sqlite3.Database(DB_PATH);
    
    return new Promise((resolve, reject) => {
        db.serialize(() => {
            // Lösche alte Einträge (außer config)
            db.run("DELETE FROM documents WHERE category != 'config'", (err) => {
                if (err) {
                    db.close();
                    return reject(err);
                }
                
                let inserted = 0;
                const stmt = db.prepare(`
                    INSERT INTO documents 
                    (name, path, category, description, type, has_symlink, symlink_path, last_update)
                    VALUES (?,?,?,?,?,?,?,?)
                `);
                
                docs.forEach(doc => {
                    stmt.run(
                        doc.name, doc.path, doc.category,
                        doc.description, doc.type, doc.has_symlink ? 1 : 0,
                        doc.symlink_path, doc.last_update
                    );
                    inserted++;
                });
                
                stmt.finalize();
                db.close();
                resolve(inserted);
            });
        });
    });
}

async function exportAll() {
    /**Erstellt alle Exporte*/
    const db = new sqlite3.Database(DB_PATH);
    
    return new Promise((resolve, reject) => {
        db.serialize(() => {
            const tables = ['documents', 'skills', 'symlinks'];
            
            let completed = 0;
            tables.forEach(table => {
                db.all(`SELECT * FROM ${table}`, (err, rows) => {
                    if (err) {
                        db.close();
                        return reject(err);
                    }
                    
                    // JSON Export
                    const jsonPath = path.join(WORKSPACE, `db_${table}.json`);
                    fs.writeFileSync(jsonPath, JSON.stringify(rows, null, 2));
                    console.log(`✅ ${jsonPath}`);
                    
                    // CSV Export
                    if (rows.length > 0) {
                        const headers = Object.keys(rows[0]);
                        const csvContent = [
                            headers.join(','),
                            ...rows.map(row => 
                                headers.map(header => `"${String(row[header]).replace(/"/g, '""')}"`).join(',')
                            )
                        ].join('\n');
                        
                        const csvPath = path.join(WORKSPACE, `db_${table}.csv`);
                        fs.writeFileSync(csvPath, csvContent);
                        console.log(`✅ ${csvPath}`);
                    }
                    
                    completed++;
                    if (completed === tables.length) {
                        db.close();
                        resolve();
                    }
                });
            });
        });
    });
}

async function main() {
    console.log("=" .repeat(60));
    console.log("DOCS.DB UPDATER");
    console.log("=" .repeat(60));
    
    console.log("\n--- Scanne Dokumentationen ---");
    const docs = scanDocumentations();
    console.log(`Gefunden: ${docs.length} Dokumente`);
    
    console.log("\n--- Aktualisiere docs.db ---");
    try {
        const inserted = await updateDatabase(docs);
        console.log(`✅ ${inserted} Dokumente in docs.db aktualisiert`);
        
        console.log("\n--- Erstelle Exporte ---");
        await exportAll();
        
        console.log("\n" + "=" .repeat(60));
        console.log("DOCS.DB AKTUALISIERT");
        console.log("=" .repeat(60));
    } catch (error) {
        console.error("Fehler beim Aktualisieren der Datenbank:", error);
    }
}

main();

File: perl5/update_docs_db.pl
#!/usr/bin/perl
# update_docs_db.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:scripts/update_docs_db.py
# Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use DBI;
use File::Find;
use File::Spec;
use JSON;
use Text::CSV;
use POSIX qw(strftime);

# Konfiguration
my $WORKSPACE = "/home/openclaw/.openclaw/workspace";
my $DB_PATH = "$WORKSPACE/db/docs.db";

# Hauptprogramm
sub main {
    print "=" x 60 . "\n";
    print "DOCS.DB UPDATER\n";
    print "=" x 60 . "\n";

    print "\n--- Scanne Dokumentationen ---\n";
    my $docs = scan_documentations();
    print "Gefunden: " . scalar(@$docs) . " Dokumente\n";

    print "\n--- Aktualisiere docs.db ---\n";
    my $inserted = update_database($docs);
    print "✅ $inserted Dokumente in docs.db aktualisiert\n";

    print "\n--- Erstelle Exporte ---\n";
    export_all();

    print "\n" . "=" x 60 . "\n";
    print "DOCS.DB AKTUALISIERT\n";
    print "=" x 60 . "\n";
}

# Scannt alle .md Dateien im Workspace
sub scan_documentations {
    my @docs;

    # Hauptverzeichnis
    opendir(my $dh, $WORKSPACE) or die "Kann $WORKSPACE nicht öffnen: $!";
    while (readdir $dh) {
        next unless /\.md$/;
        my $file = "$WORKSPACE/$_";
        next unless -f $file && !(-l $file);
        push @docs, {
            name => $_,
            path => '/',
            category => 'main',
            description => get_description($file),
            type => 'doc',
            has_symlink => 0,
            symlink_path => undef,
            last_update => get_mtime($file)
        };
    }
    closedir $dh;

    # WebSearch Verzeichnis
    my $websearch_dir = "$WORKSPACE/websearch";
    if (-d $websearch_dir) {
        opendir(my $wh, $websearch_dir) or die "Kann $websearch_dir nicht öffnen: $!";
        while (readdir $wh) {
            next unless /\.md$/;
            my $file = "$websearch_dir/$_";
            push @docs, {
                name => $_,
                path => 'websearch/',
                category => 'websearch',
                description => get_description($file),
                type => ($_ =~ /GUIDE/) ? 'guide' : 'config',
                has_symlink => 1,
                symlink_path => "websearch/$_",
                last_update => get_mtime($file)
            };
        }
        closedir $wh;
    }

    # MCP Verzeichnis
    my $mcp_dir = "$WORKSPACE/mcp";
    if (-d $mcp_dir) {
        opendir(my $mh, $mcp_dir) or die "Kann $mcp_dir nicht öffnen: $!";
        while (readdir $mh) {
            next unless /\.md$/;
            my $file = "$mcp_dir/$_";
            my $is_symlink = -l $file;
            my $symlink_path = $is_symlink ? readlink($file) : undef;
            push @docs, {
                name => $_,
                path => 'mcp/',
                category => 'mcp',
                description => get_description($file),
                type => $is_symlink ? 'symlink' : 'guide',
                has_symlink => $is_symlink,
                symlink_path => $symlink_path,
                last_update => get_mtime($file)
            };
        }
        closedir $mh;
    }

    # Docs-Unterverzeichnisse
    my $docs_dir = "$WORKSPACE/docs";
    if (-d $docs_dir) {
        opendir(my $ddh, $docs_dir) or die "Kann $docs_dir nicht öffnen: $!";
        while (my $entry = readdir $ddh) {
            next if $entry =~ /^\.\.?$/;
            my $subdir = "$docs_dir/$entry";
            next unless -d $subdir;
            opendir(my $sdh, $subdir) or die "Kann $subdir nicht öffnen: $!";
            while (my $file = readdir $sdh) {
                next unless /\.md$/;
                my $full_path = "$subdir/$file";
                push @docs, {
                    name => $file,
                    path => "docs/$entry/",
                    category => $entry,
                    description => get_description($full_path),
                    type => 'doc',
                    has_symlink => 0,
                    symlink_path => undef,
                    last_update => get_mtime($full_path)
                };
            }
            closedir $sdh;
        }
        closedir $ddh;
    }

    # Cluster, Memory, Reports, Skills
    for my $category (qw(cluster memory reports skills)) {
        my $cat_dir = "$WORKSPACE/$category";
        if (-d $cat_dir) {
            opendir(my $ch, $cat_dir) or die "Kann $cat_dir nicht öffnen: $!";
            while (my $file = readdir $ch) {
                next unless /\.md$/;
                my $full_path = "$cat_dir/$file";
                push @docs, {
                    name => $file,
                    path => "$category/",
                    category => $category,
                    description => get_description($full_path),
                    type => 'doc',
                    has_symlink => 0,
                    symlink_path => undef,
                    last_update => get_mtime($full_path)
                };
            }
            closedir $ch;
        }
    }

    return \@docs;
}

# Extrahiert erste Zeile als Beschreibung
sub get_description {
    my ($file) = @_;
    open(my $fh, '<', $file) or return 'Dokumentation';
    my $first_line = <$fh>;
    close $fh;
    chomp $first_line;
    if ($first_line =~ /^#(.*)/) {
        $first_line = $1;
        $first_line =~ s/^\s+|\s+$//g;
        return $first_line;
    }
    if (length($first_line) > 50) {
        return substr($first_line, 0, 50) . '...';
    }
    return $first_line;
}

# Gibt letzte Änderung zurück
sub get_mtime {
    my ($file) = @_;
    my @stat = stat($file);
    if (@stat) {
        my $mtime = $stat[9];
        return strftime('%Y-%m-%d', localtime($mtime));
    }
    return '2026-04-18';
}

# Aktualisiert docs.db mit allen gefundenen Dokumenten
sub update_database {
    my ($docs) = @_;
    
    my $dbh = DBI->connect("dbi:SQLite:dbname=$DB_PATH", "", "", { RaiseError => 1 }) 
        or die "Kann keine Verbindung zur Datenbank herstellen: $DBI::errstr";
    
    # Lösche alte Einträge (außer config)
    $dbh->do("DELETE FROM documents WHERE category != 'config'");
    
    # Füge neue ein
    my $inserted = 0;
    my $sth = $dbh->prepare(q{
        INSERT INTO documents 
        (name, path, category, description, type, has_symlink, symlink_path, last_update)
        VALUES (?,?,?,?,?,?,?,?)
    });
    
    for my $doc (@$docs) {
        $sth->execute(
            $doc->{name}, $doc->{path}, $doc->{category},
            $doc->{description}, $doc->{type}, $doc->{has_symlink},
            $doc->{symlink_path}, $doc->{last_update}
        );
        $inserted++;
    }
    
    $dbh->commit;
    $dbh->disconnect;
    
    return $inserted;
}

# Erstellt alle Exporte
sub export_all {
    my $dbh = DBI->connect("dbi:SQLite:dbname=$DB_PATH", "", "", { RaiseError => 1 }) 
        or die "Kann keine Verbindung zur Datenbank herstellen: $DBI::errstr";
    
    # Tabellen exportieren
    my @tables = ('documents', 'skills', 'symlinks');
    for my $table (@tables) {
        my $sth = $dbh->prepare("SELECT * FROM $table");
        $sth->execute();
        
        # Sammle Spaltennamen
        my @columns = @{$sth->{NAME}};
        
        # Sammle Daten
        my @rows;
        while (my $row = $sth->fetchrow_hashref) {
            push @rows, $row;
        }
        
        # JSON Export
        my $json_text = encode_json(\@rows);
        my $json_path = "$WORKSPACE/db_${table}.json";
        open(my $json_fh, '>', $json_path) or die "Kann $json_path nicht schreiben: $!";
        print $json_fh $json_text;
        close $json_fh;
        print "✅ $json_path\n";
        
        # CSV Export
        my $csv_path = "$WORKSPACE/db_${table}.csv";
        open(my $csv_fh, '>', $csv_path) or die "Kann $csv_path nicht schreiben: $!";
        my $csv = Text::CSV->new({ binary => 1 });
        $csv->print($csv_fh, \@columns);
        print $csv_fh "\n";
        for my $row (@rows) {
            my @values = map { $row->{$_} } @columns;
            $csv->print($csv_fh, \@values);
            print $csv_fh "\n";
        }
        close $csv_fh;
        print "✅ $csv_path\n";
    }
    
    $dbh->disconnect;
}

# Starte das Programm
main();
